### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by cohix**
Streams are a new type of handler in Atmo. A stream is a way to handle incoming streams of data (for example, a webhook connection that is continuously writing new messages). The handler's sequence will run once per message with the message as the sequence's input. For example:
```yaml
  - type: stream
    resource: /stream
    steps:
      - fn: helloworld-rs
 ```
Would create a websocket endpoint at `/stream` that runs the `helloworld-rs` Runnable for each message received over the websocket connection. The sequence can be arbitrarily complex, just like any other handler.

To test, run `make atmo bundle=./example-project/runnables.wasm.zip` and then in a seperate terminal run `go run ./testingsupport/stream-tester.go`, then write messages and press enter to see them handled by the stream handler.

### CODE REVIEW

File: atmo/coordinator/coordinator_websocket.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+	"github.com/suborbital/atmo/directive"
+	"github.com/suborbital/reactr/request"
+	"github.com/suborbital/vektor/vk"
+)
+
+func (c *Coordinator) websocketHandlerForDirectiveHandler(handler directive.Handler) http.HandlerFunc {
+	upgrader := websocket.Upgrader{} // use default options
```
Comment (jagger27) [inline atmo/coordinator/coordinator_websocket.go:15]: Should specific WebSocket upgrader options be exposed in Directive.yaml? (buffer sizes, timeouts, etc.)

File: atmo/coordinator/coordinator_websocket.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+					breakErr = err
+					break
+				}
+
+				continue
+			}
+
+			result := resultFromState(handler, seqState.state)
+
+			if err := conn.WriteMessage(websocket.TextMessage, result); err != nil {
```
Comment (jagger27) [inline atmo/coordinator/coordinator_websocket.go:72]: There should be `websocket.BinaryMessage` support

File: atmo/coordinator/coordinator_request.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func (c *Coordinator) vkHandlerForDirectiveHandler(handler directive.Handler) vk.HandlerFunc {
+	return func(r *http.Request, ctx *vk.Ctx) (interface{}, error) {
+		req, err := request.FromVKRequest(r, ctx)
+		if err != nil {
+			ctx.Log.Error(errors.Wrap(err, "failed to request.FromVKRequest"))
+			return nil, vk.E(http.StatusInternalServerError, "failed to handle request")
+		}
+
+		// this should probably be factored out into the CoordinateRequest object
```
Comment (jagger27) [inline atmo/coordinator/coordinator_request.go:21]: is this the right issue? https://github.com/suborbital/reactr/issues/105

**Review by jagger27 — APPROVED** [review]
I'm unsure about how to review vendored code. Should it be done in their own respective repos?

File: atmo/coordinator/coordinator_request.go (hunk 3 above)
Comment (cohix) [reply atmo/coordinator/coordinator_request.go:21]: @jagger27 Yes, that's right

**Comment by cohix** [comment]
> I'm unsure about how to review vendored code. Should it be done in their own respective repos?

@jagger27 If the code is vendored, that should mean it's on the vendored repo's main branch already, which means it was reviewed elsewhere. It can generally be ignored in reviews.

**Pull request merged**